File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        self.prefix = [1] * n
        self.suffix = [1] * n
        output = [1] * n

        for i in range(n-1): 
            self.prefix[i+1] = self.prefix[i] * nums[i]
        print(self.prefix)

        for i in range(n-1,0,-1):
            self.suffix[i-1] = self.suffix[i] * nums[i]
        print(self.suffix)

        for i in range(n):
            output[i] = (self.suffix[i] * self.prefix[i])
        return output

File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> output;

        for (int i=0;i<nums.size();i++) {
            if (nums[i] > 0) break;
            if (i>0 && nums[i]==nums[i-1]) continue;

            int target = -nums[i];
            int l=i+1,r=nums.size()-1;
            while (l<r) {
                if (nums[l]+nums[r] > target) {
                    r--;
                }
                else if (nums[l]+nums[r] < target) {
                    l++;
                }
                else {
                    output.push_back({nums[l],nums[i],nums[r]});
                    l++;
                    r--;
                    while (l<r && nums[l] == nums[l-1]) l++;
                }
            }
        }
        return output;
    }
};
